bigquery_etl/deploy.py

```diff
@@ -0,0 +1,120 @@
+
+
+class FailedDeployException(Exception):
+    """Raised for failed deployments."""
+
+
+def deploy_table(
+    query_file: Path,
+    destination_table: Optional[str] = None,
+    force: bool = False,
+    use_cloud_function: bool = False,
+    skip_existing: bool = False,
+    update_metadata: bool = True,
+    respect_dryrun_skip: bool = True,
+    sql_dir=ConfigLoader.get("default", "sql_dir"),
+) -> None:
+    """Deploy a query to a destination."""
+    if respect_dryrun_skip and str(query_file) in DryRun.skipped_files():
+        raise SkippedDeployException(f"Dry run skipped for {query_file}.")
+
+    try:
+        metadata = Metadata.of_query_file(query_file)
+        if (
+            metadata.scheduling
+            and "destination_table" in metadata.scheduling
+            and metadata.scheduling["destination_table"] is None
+        ):
+            raise FailedDeployException(
+                f"Destination table configured for {query_file} in metadata but is empty."
+            )
+    except FileNotFoundError:
+        log.warning(f"No metadata found for {query_file}.")
+
+    table_name = query_file.parent.name
+    dataset_name = query_file.parent.parent.name
+    project_name = query_file.parent.parent.parent.name
+
+    if destination_table is None:
+        destination_table = f"{project_name}.{dataset_name}.{table_name}"
+
+    existing_schema_path = query_file.parent / SCHEMA_FILE
+    try:
+        existing_schema = Schema.from_schema_file(existing_schema_path)
+    except Exception as e:  # TODO: Raise/catch more specific exception
+        raise SkippedDeployException(f"Schema missing for {query_file}.") from e
+
+    if not force and str(query_file).endswith("query.sql"):
+        query_schema = Schema.from_query_file(
+            query_file,
+            use_cloud_function=use_cloud_function,
+            respect_skip=respect_dryrun_skip,
+            sql_dir=sql_dir,
+        )
+        if not existing_schema.equal(query_schema):
+            raise FailedDeployException(
+                f"Query {query_file} does not match "
+                f"schema in {existing_schema_path}. "
+                f"To update the local schema file, "
+                f"run `./bqetl query schema update "
+                f"{dataset_name}.{table_name}`",
+            )
+
+    client = bigquery.Client()
+    try:
+        table = client.get_table(destination_table)
+    except NotFound:
+        table = bigquery.Table(destination_table)
+    table.schema = existing_schema.to_bigquery_schema()
+
+    if update_metadata:
+        attach_metadata(query_file, table)
+
+    _create_or_update(client, table, destination_table, skip_existing)
+
+
+def _create_or_update(
+    client: bigquery.Client,
+    table: bigquery.Table,
+    destination_table: str,
+    skip_existing: bool = False,
+) -> None:
+    if table.created:
+        if skip_existing:
+            raise SkippedDeployException(f"{destination_table} already exists.")
```

Comment:
I wanted to use `table.full_table_id` here, but this apparently doesn't get populated when you create a `google.cloud.bigquery.Table` object 🤔 Assigning it (similar to the schema/metadata) seems strange given we already provide a table reference str in the constructor. Curious for thoughts here if any.

Reply:
I believe `table.table_id` and `table.dataset_id` should be available.

The [docs](https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.table.Table#google_cloud_bigquery_table_Table_full_table_id) mention: _None until set from the server_

Reply:
`str(table)` actually gives the full id as well https://github.com/googleapis/python-bigquery/blob/f19d39801cb5828929f1117ab93492aefdada5fd/google/cloud/bigquery/table.py#L1092-L1093

Reply:
Ahh good find! I'll use the string representation